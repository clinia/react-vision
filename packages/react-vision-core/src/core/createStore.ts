type SearchState = any;
type ResultsState = any;
type SuggestionsState = any;
type Listener = () => void;
type State = {
  widgets: SearchState;
  metadata: any[];
  results: ResultsState | null;
  resultsSuggestions: SuggestionsState | null;
  error: Error | null;
  searching: boolean;
  searchingForSuggestions: boolean;
  isSearchStalled: boolean;
};
export default function createStore(initialState: State) {
  let state = initialState;
  const listeners: Listener[] = [];
  return {
    getState() {
      return state;
    },
    setState(nextState: State) {
      state = nextState;
      listeners.forEach(listener => listener());
    },
    subscribe(listener: Listener) {
      listeners.push(listener);
      return function unsubscribe() {
        listeners.splice(listeners.indexOf(listener), 1);
      };
    },
  };
}

export type Store = ReturnType<typeof createStore>;
