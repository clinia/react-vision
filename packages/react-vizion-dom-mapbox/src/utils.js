import PropTypes from 'prop-types';

export const registerEvents = (events, props, instance) => {
  const eventsAvailable = Object.keys(events);
  const keys = Object.keys(props).filter(
    key => eventsAvailable.indexOf(key) !== -1
  );
  keys.forEach(name =>
    instance.addEventListener(events[name], event => {
      props[name]({ event, marker: instance });
    })
  );

  return () => {
    keys.forEach(name =>
      instance.removeEventListener(events[name], event => {
        props[name]({ event, marker: instance });
      })
    );
  };
};

export const createListenersPropTypes = eventTypes =>
  Object.keys(eventTypes).reduce(
    (acc, name) => ({ ...acc, [name]: PropTypes.func }),
    {}
  );

export const createFilterProps = excludes => props =>
  Object.keys(props)
    .filter(name => excludes.indexOf(name) === -1)
    .reduce(
      (acc, name) => ({
        ...acc,
        [name]: props[name],
      }),
      {}
    );

export const betweenRange = (value, n1, n2) => {
  return n1 <= value && n2 >= value;
};

export const isValidCoordinates = (lat, lon) => {
  return latitude(lat) && longitude(lon);
};

const isNumber = n => {
  return typeof n === 'number';
};

function longitude(lon) {
  return Boolean(isNumber(lon) && betweenRange(lon, -180, 180));
}

function latitude(lat) {
  return Boolean(isNumber(lat) && betweenRange(lat, -90, 90));
}
