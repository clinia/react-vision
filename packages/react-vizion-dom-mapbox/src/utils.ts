export const betweenRange = (value: number, n1: number, n2: number) => {
  return n1 <= value && n2 >= value;
};

export const isValidCoordinates = (lat: any, lon: any) => {
  return latitude(lat) && longitude(lon);
};

const isNumber = (n: any) => {
  return typeof n === 'number';
};

function longitude(lon: any) {
  return Boolean(isNumber(lon) && betweenRange(lon, -180, 180));
}

function latitude(lat: any) {
  return Boolean(isNumber(lat) && betweenRange(lat, -90, 90));
}
