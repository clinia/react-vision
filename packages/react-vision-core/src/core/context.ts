import { createContext } from 'react';
import { Store } from '../core/createStore';

export type VisionContext = {
  onInternalStateUpdate: (...args: any[]) => any;
  createHrefForState: (...args: any[]) => string;
  onSearchForSuggestions: (...args: any[]) => any;
  onSearchStateChange: (...args: any[]) => any;
  onSearchParameters: (...args: any[]) => any;
  store: Store;
  widgetsManager: any;
  mainTargetedIndex: string;
};

export const {
  Consumer: VisionConsumer,
  Provider: VisionProvider,
} = createContext<VisionContext>({
  onInternalStateUpdate: () => undefined,
  createHrefForState: () => '#',
  onSearchForSuggestions: () => undefined,
  onSearchStateChange: () => undefined,
  onSearchParameters: () => undefined,
  store: {} as Store,
  widgetsManager: {},
  mainTargetedIndex: '',
});

export type IndexContext =
  | {
      targetedIndex: string;
    }
  | undefined;

export const {
  Consumer: IndexConsumer,
  Provider: IndexProvider,
} = createContext<IndexContext>(undefined);
