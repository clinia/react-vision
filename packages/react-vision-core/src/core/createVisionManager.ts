import cliniasearchHelper from 'cliniasearch-helper';
import createWidgetsManager from './createWidgetsManager';
import createStore from './createStore';
import { HIGHLIGHT_TAGS } from './highlight';
import { hasMultipleIndices } from './indexUtils';
import { version as ReactVersion } from 'react';
import version from './version';

function addCliniaAgents(searchClient) {
  if (typeof searchClient.addCliniaAgent === 'function') {
    searchClient.addCliniaAgent(`react (${ReactVersion})`);
    searchClient.addCliniaAgent(`react-vision (${version})`);
  }
}

const isMultiIndexContext = widget =>
  hasMultipleIndices({
    vci: widget.props.contextValue,
    multiIndexContext: widget.props.indexContextValue,
  });
const isTargetedIndexEqualIndex = (widget, indexId) =>
  widget.props.indexContextValue.targetedIndex === indexId;

const isIndexWidget = widget => Boolean(widget.props.indexId);
const isIndexWidgetEqualIndex = (widget, indexId) =>
  widget.props.indexId === indexId;

const sortIndexWidgetsFirst = (firstWidget, secondWidget) => {
  if (isIndexWidget(firstWidget)) {
    return -1;
  }
  if (isIndexWidget(secondWidget)) {
    return 1;
  }
  return 0;
};

/**
 * Creates a new instance of the VisionManager which controls the widgets and
 * trigger the search when the widgets are updated.
 * @param {string} indexName - the main index name
 * @param {object} initialState - initial widget state
 * @param {object} SearchParameters - optional additional parameters to send to the clinia API
 * @param {number} stalledSearchDelay - time (in ms) after the search is stalled
 * @return {VisionManager} a new instance of VisionManager
 */
export default function createVisionManager({
  indexName,
  initialState = {},
  searchClient,
  resultsState,
  stalledSearchDelay,
}) {
  const helper = cliniasearchHelper(searchClient, indexName, {
    ...HIGHLIGHT_TAGS,
  });

  addCliniaAgents(searchClient);

  helper
    .on('search', handleNewSearch)
    .on('result', handleSearchSuccess({ indexId: indexName }))
    .on('error', handleSearchError);

  let locationClient = searchClient.initPlaces();

  let skip = false;
  let stalledSearchTimer: NodeJS.Timeout | null = null;
  let initialSearchParameters = helper.state;

  const widgetsManager = createWidgetsManager(onWidgetsUpdate);

  hydrateSearchClient(searchClient, resultsState);

  const store = createStore({
    widgets: initialState,
    metadata: [],
    results: hydrateResultsState(resultsState),
    resultsSuggestions: null,
    resultsLocations: null,
    error: null,
    searching: false,
    searchingForSuggestions: false,
    searchingForLocations: false,
    isSearchStalled: true,
  });

  function skipSearch() {
    skip = true;
  }

  function updateClient(client) {
    addCliniaAgents(client);
    helper.setClient(client);
    search();
  }

  function clearCache() {
    helper.clearCache();
    search();
  }

  function getMetadata(state) {
    return widgetsManager
      .getWidgets()
      .filter(widget => Boolean(widget.getMetadata))
      .map(widget => widget.getMetadata(state));
  }

  function getSearchParameters() {
    const sharedParameters = widgetsManager
      .getWidgets()
      .filter(widget => Boolean(widget.getSearchParameters))
      .filter(widget => !isMultiIndexContext(widget) && !isIndexWidget(widget))
      .reduce(
        (res, widget) => widget.getSearchParameters(res),
        initialSearchParameters
      );

    const mainParameters = widgetsManager
      .getWidgets()
      .filter(widget => Boolean(widget.getSearchParameters))
      .filter(widget => {
        const targetedIndexEqualMainIndex =
          isMultiIndexContext(widget) &&
          isTargetedIndexEqualIndex(widget, indexName);

        const subIndexEqualMainIndex =
          isIndexWidget(widget) && isIndexWidgetEqualIndex(widget, indexName);

        return targetedIndexEqualMainIndex || subIndexEqualMainIndex;
      })
      .sort(sortIndexWidgetsFirst)
      .reduce(
        (res, widget) => widget.getSearchParameters(res),
        sharedParameters
      );

    const derivedIndices = widgetsManager
      .getWidgets()
      .filter(widget => Boolean(widget.getSearchParameters))
      .filter(widget => {
        const targetedIndexNotEqualMainIndex =
          isMultiIndexContext(widget) &&
          !isTargetedIndexEqualIndex(widget, indexName);

        const subIndexNotEqualMainIndex =
          isIndexWidget(widget) && !isIndexWidgetEqualIndex(widget, indexName);

        return targetedIndexNotEqualMainIndex || subIndexNotEqualMainIndex;
      })
      .sort(sortIndexWidgetsFirst)
      .reduce((indices, widget) => {
        const indexId = isMultiIndexContext(widget)
          ? widget.props.indexContextValue.targetedIndex
          : widget.props.indexId;

        const widgets = indices[indexId] || [];

        return { ...indices, [indexId]: widgets.concat(widget) };
      }, {});

    const derivedParameters = Object.keys(derivedIndices).map(indexId => ({
      parameters: derivedIndices[indexId].reduce(
        (res, widget) => widget.getSearchParameters(res),
        sharedParameters
      ),
      indexId,
    }));

    return { mainParameters, derivedParameters };
  }

  function onSearchForSuggestions({ query, args }) {
    store.setState({
      ...store.getState(),
      searchingForSuggestions: true,
    });

    searchClient
      .suggest(query, args)
      .then(content => {
        store.setState({
          ...store.getState(),
          searchingForSuggestions: false,
          resultsSuggestions: {
            suggestions: content,
            query,
          },
        });
      })
      .catch(error => {
        // Since setState is synchronous, any error that occurs in the render of a
        // component will be swallowed by this promise.
        // This is a trick to make the error show up correctly in the console.
        // See http://stackoverflow.com/a/30741722/969302
        setTimeout(() => {
          throw error;
        });
      });
  }

  function onSearchForLocations({ query, ...args }) {
    store.setState({
      ...store.getState(),
      searchingForLocations: true,
    });

    if (locationClient) {
      locationClient!
        .suggest(query, args)
        .then(content => {
          store.setState({
            ...store.getState(),
            searchingForLocations: false,
            resultsLocations: {
              suggestions: content,
              query,
            },
          });
        })
        .catch(error => {
          // Since setState is synchronous, any error that occurs in the render of a
          // component will be swallowed by this promise.
          // This is a trick to make the error show up correctly in the console.
          // See http://stackoverflow.com/a/30741722/969302
          setTimeout(() => {
            throw error;
          });
        });
    }
  }

  function search() {
    if (skip) return;

    const { mainParameters, derivedParameters } = getSearchParameters();

    helper.derivedHelpers.slice().forEach(derivedHelper => {
      derivedHelper.detach();
    });

    derivedParameters.forEach(({ indexId, parameters }) => {
      const derivedHelper = helper.derive(() => parameters);

      derivedHelper
        .on('result', handleSearchSuccess({ indexId }))
        .on('error', handleSearchError);
    });

    helper.setState(mainParameters);

    helper.search();
  }

  function handleSearchSuccess({ indexId }) {
    return event => {
      const state = store.getState();
      const isDerivedHelperEmpty = !helper.derivedHelpers.length;

      let results = state.results ? state.results : {};

      if (!isDerivedHelperEmpty) {
        results[indexId] = event;
      } else {
        results = event;
      }

      const currentState = store.getState();
      let nextIsSearchStalled = currentState.isSearchStalled;
      if (!helper.hasPendingRequests()) {
        clearTimeout(stalledSearchTimer!);
        stalledSearchTimer = null;
        nextIsSearchStalled = false;
      }

      const { ...partialState } = currentState;

      store.setState({
        ...partialState,
        results,
        isSearchStalled: nextIsSearchStalled,
        searching: false,
        error: null,
      });
    };
  }

  function handleSearchError({ error }) {
    const currentState = store.getState();

    let nextIsSearchStalled = currentState.isSearchStalled;
    if (!helper.hasPendingRequests()) {
      clearTimeout(stalledSearchTimer!);
      nextIsSearchStalled = false;
    }

    const { ...partialState } = currentState;

    store.setState({
      ...partialState,
      isSearchStalled: nextIsSearchStalled,
      error,
      searching: false,
    });
  }

  function handleNewSearch() {
    if (stalledSearchTimer) return;

    stalledSearchTimer = setTimeout(() => {
      const { ...partialState } = store.getState();

      store.setState({
        ...partialState,
        isSearchStalled: true,
      });
    }, stalledSearchDelay);
  }

  function hydrateSearchClient(client, results) {
    if (!results) return;

    if (!client._useCache || typeof client.addCliniaAgent !== 'function') {
      // Avoid hydrating the client when cache is disable or the client is not ours
      return;
    }

    if (Array.isArray(results)) {
      hydrateSearchClientWithMultiIndexRequest(client, results);
      return;
    }

    hydrateSearchClientWithSingleIndexRequest(client, results);
  }

  function hydrateSearchClientWithMultiIndexRequest(client, results) {
    const key = `/1/indexes/*/queries_body_${JSON.stringify({
      requests: results.reduce((acc, result) =>
        acc.concat(
          result.rawResults.map(request => ({
            indexName: request.index,
            params: request.params,
          }))
        )
      ),
    })}`;

    client.cache = {
      ...client.cache,
      [key]: JSON.stringify({
        results: results.reduce(
          (acc, result) => acc.concat(result.rawResults),
          []
        ),
      }),
    };
  }

  function hydrateSearchClientWithSingleIndexRequest(client, results) {
    const key = `/1/indexes/*/queries_body_${JSON.stringify({
      requests: results.rawResults.map(request => ({
        indexName: request.index,
        params: request.params,
      })),
    })}`;

    client.cache = {
      ...client.cache,
      [key]: JSON.stringify({
        results: results.rawResults,
      }),
    };
  }

  function hydrateResultsState(results) {
    if (!results) return null;

    if (Array.isArray(results)) {
      return results.reduce(
        (acc, result) => ({
          ...acc,
          [result._internalIndexId]: new cliniasearchHelper.SearchResults(
            new cliniasearchHelper.SearchParameters(result.state),
            result.rawResults
          ),
        }),
        {}
      );
    }

    return new cliniasearchHelper.SearchResults(
      new cliniasearchHelper.SearchParameters(results.state),
      results.rawResults
    );
  }

  // Called whenever a widget has been rendered with new props.
  function onWidgetsUpdate() {
    const metadata = getMetadata(store.getState().widgets);

    store.setState({
      ...store.getState(),
      metadata,
      searching: true,
    });

    // Since the `getSearchParameters` method of widgets also depends on props,
    // the result search parameters might have changed.
    search();
  }

  function transitionState(nextSearchState) {
    const searchState = store.getState().widgets;

    return widgetsManager
      .getWidgets()
      .filter(widget => Boolean(widget.transitionState))
      .reduce(
        (res, widget) => widget.transitionState(searchState, res),
        nextSearchState
      );
  }

  function onExternalStateUpdate(nextSearchState) {
    const metadata = getMetadata(nextSearchState);

    store.setState({
      ...store.getState(),
      widgets: nextSearchState,
      metadata,
      searching: true,
    });

    search();
  }

  function updateIndex(newIndex) {
    initialSearchParameters = initialSearchParameters.setIndex(newIndex);
    // No need to trigger a new search here as the widgets will also update and trigger it if needed.
  }

  function getWidgetsIds() {
    return store
      .getState()
      .metadata.reduce(
        (res, meta) =>
          typeof meta.id !== 'undefined' ? res.concat(meta.id) : res,
        []
      );
  }

  return {
    store,
    widgetsManager,
    getWidgetsIds,
    getSearchParameters,
    onSearchForSuggestions,
    onSearchForLocations,
    onExternalStateUpdate,
    transitionState,
    updateClient,
    updateIndex,
    clearCache,
    skipSearch,
  };
}

export type VisionManager = ReturnType<typeof createVisionManager>;
