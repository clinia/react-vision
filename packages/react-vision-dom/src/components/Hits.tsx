import React from 'react';
import classnames from 'classnames';
import { createClassNames } from '../core/utils';
import { translatable } from 'react-vision-core';

export type SearchResult = {
  id: string;
};

type HitProps = {
  results: SearchResult[];
  className?: string;
  style?: React.CSSProperties;
  noResultsFound?: React.ReactNode;
  hit?: (result: SearchResult) => React.ReactNode;
  onClick?: (event?: React.MouseEvent) => void;
  translate: any;
};

type DefaultHitProps = {
  searchResult: SearchResult;
};

const DefaultHit: React.FunctionComponent<DefaultHitProps> = ({
  searchResult,
}) => {
  return (
    <div className="default-hit-card">
      <div className="default-hit-container">
        <h4 className="default-hit-title">{searchResult.id}</h4>
        <div className="card-body">{JSON.stringify(searchResult, null, 2)}</div>
      </div>
    </div>
  );
};

const cx = createClassNames('Hits');

const Hits: React.FunctionComponent<HitProps> = ({
  results,
  className,
  style,
  hit,
  noResultsFound,
  onClick,
  translate,
}) => {
  return (
    <div className={classnames(cx('hits'), className)} style={style}>
      <div className={cx('hits-list')}>
        {Array.isArray(results) && results.length > 0
          ? results.map(searchResult => (
              <div key={searchResult.id} onClick={onClick}>
                {hit ? (
                  hit(searchResult)
                ) : (
                  <DefaultHit searchResult={searchResult} />
                )}
              </div>
            ))
          : noResultsFound || (
              //Temporarily until we have the styles decided
              <div className={cx('empty')}>{translate('emptySearch')}</div>
            )}
      </div>
    </div>
  );
};

export default translatable({
  emptySearch: 'No results were found',
})(Hits);
