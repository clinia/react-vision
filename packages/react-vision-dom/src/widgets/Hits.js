import { connectHits } from 'react-vision-core';
import Hits from '../components/Hits';

export default connectHits(Hits);
