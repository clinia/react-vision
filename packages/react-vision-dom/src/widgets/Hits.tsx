import { connectHits } from 'react-vision-core';
import Hits from '../components/Hits';

/**
 * Displays a list of hits.
 *
 * @name Hits
 * @kind widget
 * @propType {Component} [hitComponent] - Component used for rendering each hit from
 *   the results. If it is not provided the rendering defaults to displaying the
 *   hit in its JSON form. The component will be called with a `record` prop.
 * @propType {node} [noResultsFound] - Change the apparence of the default empty results found.
 * @propType {string} [className] - Add a custom CSS class to the component
 * @propType {object} [style] - Add a custom React.CSSProperties object to the Hits container
 * @themeKey cvi-Hits - the root div of the widget
 * @themeKey cvi-Hits-list - the list of results
 * @themeKey cvi-Hits-item - the hit list item
 * @translationkey emptySearch - The text shown on empty results
 * @example
 * import React from 'react';
 * import cliniasearch from 'cliniasearch/lite';
 * import { Vision, Hits } from 'react-vision-dom';
 *
 * const searchClient = cliniasearch(
 *   'TODO',
 *   'test'
 * );
 * const App = () => (
 *   <Vision
 *     searchClient={searchClient}
 *     indexName="health_facility"
 *   >
 *     <Hits />
 *   </Vision>
 * );
 */

export default connectHits(Hits);
