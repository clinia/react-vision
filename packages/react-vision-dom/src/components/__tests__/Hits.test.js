import React from 'react';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import Hits from '../Hits';
import Enzyme, { shallow, mount } from 'enzyme';
import { createClassNames } from '../../core/utils';

Enzyme.configure({ adapter: new Adapter() });

/**
 * NOTES:
 * Tests that uses .dive() do that to get the inner component inside the hoc
 * To our tests it's necessary to retrieve the inner component to simulate events and etc
 */
describe('Hits', () => {
  const cx = createClassNames('Hits');
  it('applies its default props', () => {
    const instance = renderer.create(<Hits refine={() => null} />);

    expect(instance.toJSON()).toMatchSnapshot();

    instance.unmount();
  });

  it('should render with default hit component', () => {
    const results = [{ id: '123', name: 'test' }];
    const wrapper = shallow(<Hits results={results} />);

    expect(wrapper.exists('.default-hit-card')).toBeFalsy();
  });

  it('should render with default no results found', () => {
    const wrapper = mount(<Hits />);
    expect(wrapper.exists(`.${cx('empty')}`)).toBeTruthy();
  });

  describe('Style tests', () => {
    it('should render main div with className props ', () => {
      const classNameTest = 'class-test';
      const wrapper = shallow(<Hits className={classNameTest} />);

      expect(wrapper.first('div').hasClass(classNameTest)).toBeTruthy();
    });

    it('should render main div with style props ', () => {
      const styleTest = { top: 0 };
      const wrapper = shallow(<Hits style={styleTest} />);

      expect(wrapper.first('div').props().style).toEqual(styleTest);
    });
  });

  describe('Custom inner components', () => {
    it('should render with custom hit component', () => {
      const records = [{ id: '123', name: 'test' }];
      const customHitComponent = <h1>Test</h1>;
      const wrapper = shallow(
        <Hits records={records} hit={_ => customHitComponent} />
      ).dive();

      expect(wrapper.contains(customHitComponent)).toBeTruthy();
      expect(wrapper.exists('.default-hit-no-results-found')).toBeFalsy();
    });

    it('should render with custom noResultsFound component', () => {
      const customNoResultsFound = <h1>Test</h1>;
      const wrapper = shallow(
        <Hits noResultsFound={customNoResultsFound} />
      ).dive();

      expect(wrapper.contains(customNoResultsFound)).toBeTruthy();
    });
  });

  describe('DefaultHit', () => {
    it('should render with DefaultHit with resource id in the title', () => {
      const records = [{ id: '123', name: 'test' }];
      const wrapper = mount(<Hits records={records} />);
      const titleElement = wrapper.find('h4');

      expect(titleElement.text()).toEqual(records[0].id);
    });

    it('should render DefaultHit with the resource as a JSON in the card body', () => {
      const records = [{ id: '123', name: 'test' }];
      const wrapper = mount(<Hits records={records} />);

      const cardBodyElement = wrapper.find('code');

      const expectedResult = JSON.stringify(records[0], null, 2);

      expect(cardBodyElement.text()).toEqual(expectedResult);
    });
  });
});
