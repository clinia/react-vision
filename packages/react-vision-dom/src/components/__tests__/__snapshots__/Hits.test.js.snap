// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hits applies its default props 1`] = `
<div
  className="cvi-Hits-hits"
>
  <div
    className="cvi-Hits-hits-list"
  >
    <div
      className="cvi-Hits-empty"
    >
      No results were found
    </div>
  </div>
</div>
`;
